### ISSUE TITLE
(3.0.0-M2) Bug/New Feature 559307 change needs doPrivilege() with TreadMXBean.getThreadInfo() call

### ISSUE THREAD

**Issue by jgrassel** [eclipse-ee4j/eclipselink#972]
While testing with the new M2 build of Eclipselink, I encountered an unexpected AccessControlException:

```
("java.lang.management.ManagementPermission" "monitor")
Stack: 
java.security.AccessControlException: access denied ("java.lang.management.ManagementPermission" "monitor")java.security.AccessControlContext.checkPermission(AccessControlContext.java:472)
java.security.AccessController.checkPermission(AccessController.java:884)
java.lang.SecurityManager.checkPermission(SecurityManager.java:549)
com.ibm.ws.kernel.launch.internal.MissingDoPrivDetectionSecurityManager.checkPermission(MissingDoPrivDetectionSecurityManager.java:45)
sun.management.Util.checkAccess(Util.java:77)
sun.management.Util.checkMonitorAccess(Util.java:82)
sun.management.ThreadImpl.getThreadInfo(ThreadImpl.java:172)
org.eclipse.persistence.internal.helper.ConcurrencyUtil.enrichGenerateThreadDumpForCurrentThread(ConcurrencyUtil.java:797)
org.eclipse.persistence.internal.helper.ConcurrencyUtil.createReadLockAcquisitionMetadata(ConcurrencyUtil.java:1468)
org.eclipse.persistence.internal.helper.ReadLockManager.addReadLock(ReadLockManager.java:49)
org.eclipse.persistence.internal.helper.ConcurrencyManager.addReadLockToReadLockManager(ConcurrencyManager.java:817)
org.eclipse.persistence.internal.helper.ConcurrencyManager.acquireReadLock(ConcurrencyManager.java:341)
org.eclipse.persistence.internal.identitymaps.CacheKey.acquireReadLock(CacheKey.java:288)
org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.cloneAndRegisterObject(UnitOfWorkImpl.java:1052)
org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.cloneAndRegisterObject(UnitOfWorkImpl.java:1007)
org.eclipse.persistence.internal.sessions.UnitOfWorkIdentityMapAccessor.getAndCloneCacheKeyFromParent(UnitOfWorkIdentityMapAccessor.java:212)
org.eclipse.persistence.internal.sessions.UnitOfWorkIdentityMapAccessor.getFromIdentityMap(UnitOfWorkIdentityMapAccessor.java:140)
org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.registerExistingObject(UnitOfWorkImpl.java:4042)
org.eclipse.persistence.queries.ObjectBuildingQuery.registerIndividualResult(ObjectBuildingQuery.java:461)
org.eclipse.persistence.queries.ReadObjectQuery.registerResultInUnitOfWork(ReadObjectQuery.java:898)
org.eclipse.persistence.queries.ObjectLevelReadQuery.checkEarlyReturn(ObjectLevelReadQuery.java:949)
org.eclipse.persistence.queries.DatabaseQuery.execute(DatabaseQuery.java:849)
org.eclipse.persistence.queries.ObjectLevelReadQuery.execute(ObjectLevelReadQuery.java:1195)
org.eclipse.persistence.queries.ReadObjectQuery.execute(ReadObjectQuery.java:448)
org.eclipse.persistence.queries.ObjectLevelReadQuery.executeInUnitOfWork(ObjectLevelReadQuery.java:1283)
org.eclipse.persistence.internal.sessions.UnitOfWorkImpl.internalExecuteQuery(UnitOfWorkImpl.java:2984)
org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1844)
org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1826)
org.eclipse.persistence.internal.sessions.AbstractSession.executeQuery(AbstractSession.java:1776)
org.eclipse.persistence.internal.jpa.EntityManagerImpl.executeQuery(EntityManagerImpl.java:1012)
org.eclipse.persistence.internal.jpa.EntityManagerImpl.findInternal(EntityManagerImpl.java:954)
org.eclipse.persistence.internal.jpa.EntityManagerImpl.find(EntityManagerImpl.java:830)
org.eclipse.persistence.internal.jpa.EntityManagerImpl.find(EntityManagerImpl.java:696)
```

It looks like this was added with `Bug/New Feature 559307: EclipseLink on all versions can dead-lock forever (#927)` (commit 9f4ec056c1b361f9c9a2530bb892c5fbd0df28e5).

The affected code:
```
       try {
            // (a) search for the stack trace of the current
            final StringWriter writer = new StringWriter();
            final ThreadMXBean threadMXBean = ManagementFactory.getThreadMXBean();
// The following line needs to be in a doPriv() block
            final ThreadInfo[] threadInfos = threadMXBean.getThreadInfo(new long[] { currentThreadId }, 700); 
            for (ThreadInfo threadInfo : threadInfos) {
                enrichGenerateThreadDumpForThreadInfo(writer, threadInfo);
            }
            return writer.toString();
        } catch (Exception failToAcquireThreadDumpProgrammatically) {
            AbstractSessionLog.getLog().logThrowable(SessionLog.SEVERE, SessionLog.CACHE, failToAcquireThreadDumpProgrammatically);
            return TraceLocalization.buildMessage("concurrency_util_enrich_thread_dump", new Object[] {failToAcquireThreadDumpProgrammatically.getMessage()});
        }
    }
```

**Comment by rfelcman** [comment]
Could You please provide some details which leads in this bug like JDK version/provider, test code which leads into error, test environment (JSE or JEE), maybe security settings passed to Java.

**Comment by lukasj** [comment]
@rfelcman It should be enough to enable securitymanager for tests (server-side ones)

**Comment by jgrassel** [comment]
Like Lukas said, turning on security manager would be enough.  The javadoc states that the method ThreadMXBean.getThreadInfo() requires the ("java.lang.management.ManagementPermission" "monitor") privilege.  

While my test bombed with the above, I did notice that there are other uses of ThreadMXBean methods being used that are also protected, like getAllThreadIds().  Those also have to be doPriv()'d, so the example I posted above is not the only one.
